In [1]:
import numpy as np

# Specify the path to your CSV file
csv_file_path = '/Users/noahjillson/Desktop/MCM/Wimbledon_featured_matches.csv'

# Load data from CSV file into a NumPy array
data_array = np.genfromtxt(csv_file_path, delimiter=',',skip_header=1)

# Print the NumPy array
print(data_array)

[[nan nan nan ... nan nan nan]
 [nan nan nan ... nan nan nan]
 [nan nan nan ... nan nan nan]
 ...
 [nan nan nan ... nan nan nan]
 [nan nan nan ... nan nan nan]
 [nan nan nan ... nan nan nan]]


In [21]:
import csv
# Specify the path to your CSV file
csv_file_path = '/Users/noahjillson/Desktop/MCM/Wimbledon_featured_matches.csv'

# Open the CSV file in read mode
alcaraz_interactions = []
with open(csv_file_path, 'r') as file:
    # Create a CSV reader object that returns rows as dictionaries
    csv_reader = csv.DictReader(file)

    # Iterate through the rows in the CSV file
    for row in csv_reader:
        # Each row is a dictionary with keys as header names
        if (row['player1'] == "Carlos Alcaraz" or row['player2'] == "Carlos Alcaraz"):
            alcaraz_interactions.append(row)

In [22]:
np_alcaraz = np.array(alcaraz_interactions)

# Calculating Probability Of Winning Next Point (Given Previous Point Won)

### Dictionary of all the games Alcaraz played and his opponent

In [25]:
a_games = {}
players = set()
for r in np_alcaraz:
    players.add((r['player1'], r['player2']))
    if r['player1'] == "Carlos Alcaraz":
        a_games[r['match_id']] = r['player2']
    else:
        a_games[r['match_id']] = r['player1']

print(a_games)
print(len(a_games))
print(players)

{'2023-wimbledon-1301': 'Nicolas Jarry', '2023-wimbledon-1401': 'Matteo Berrettini', '2023-wimbledon-1501': 'Holger Rune', '2023-wimbledon-1601': 'Daniil Medvedev', '2023-wimbledon-1701': 'Novak Djokovic'}
5
{('Carlos Alcaraz', 'Daniil Medvedev'), ('Carlos Alcaraz', 'Novak Djokovic'), ('Carlos Alcaraz', 'Holger Rune'), ('Carlos Alcaraz', 'Nicolas Jarry'), ('Carlos Alcaraz', 'Matteo Berrettini')}


In [29]:
matches = {}

for m in a_games:
    matches[m] = list()

for m in a_games:
    for r in np_alcaraz:
        if r['match_id'] == m:
            matches[m].append(r)

print(matches['2023-wimbledon-1301'])

[{'match_id': '2023-wimbledon-1301', 'player1': 'Carlos Alcaraz', 'player2': 'Nicolas Jarry', 'elapsed_time': '00:00:00', 'set_no': '1', 'game_no': '1', 'point_no': '1', 'p1_sets': '0', 'p2_sets': '0', 'p1_games': '0', 'p2_games': '0', 'p1_score': '0', 'p2_score': '0', 'server': '1', 'serve_no': '2', 'point_victor': '2', 'p1_points_won': '0', 'p2_points_won': '1', 'game_victor': '0', 'set_victor': '0', 'p1_ace': '0', 'p2_ace': '0', 'p1_winner': '0', 'p2_winner': '0', 'winner_shot_type': '0', 'p1_double_fault': '0', 'p2_double_fault': '0', 'p1_unf_err': '1', 'p2_unf_err': '0', 'p1_net_pt': '0', 'p2_net_pt': '0', 'p1_net_pt_won': '0', 'p2_net_pt_won': '0', 'p1_break_pt': '0', 'p2_break_pt': '0', 'p1_break_pt_won': '0', 'p2_break_pt_won': '0', 'p1_break_pt_missed': '0', 'p2_break_pt_missed': '0', 'p1_distance_run': '6', 'p2_distance_run': '7.84', 'rally_count': '2', 'speed_mph': '95', 'serve_width': 'BC', 'serve_depth': 'NCTL', 'return_depth': 'ND'}, {'match_id': '2023-wimbledon-1301', 'p

In [39]:
p_sum = 0
p_consec_sum = 0
for mr in matches.values():
    pl = None
    cpnum = '2'
    if mr[0]['player1'] == 'Carlos Alcaraz':
        cpnum = '1'
    for r in mr:
        if pl is None:
            pl = r['point_victor']
            continue
        # If Calros won last point
        if cpnum == pl:
            # If Carlos won current point
            if r['point_victor'] == cpnum:
                # inc sum to indicate carlos won current point after winning the previous one
                p_consec_sum += 1 
            # Increment sum of opportunities for carlso to win the next point
            p_sum += 1

        pl = r['point_victor']

print(p_sum, p_consec_sum, p_consec_sum/p_sum)

637 347 0.5447409733124019


In [40]:
p_sum = 0
p_consec_sum = 0
for mr in matches.values():
    pl = None
    cpnum = '2'
    if mr[0]['player1'] == 'Carlos Alcaraz':
        cpnum = '1'
    for r in mr:
        if pl is None:
            pl = r['point_victor']
            continue
        # If Calros lost last point
        if cpnum != pl:
            # If Carlos won current point
            if r['point_victor'] == cpnum:
                # inc sum to indicate carlos won current point after winning the previous one
                p_consec_sum += 1 
            # Increment sum of opportunities for carlso to win the next point
            p_sum += 1

        pl = r['point_victor']

print(p_sum, p_consec_sum, p_consec_sum/p_sum)

565 292 0.5168141592920354


In [41]:
p_sum = 0
p_consec_sum = 0
for mr in matches.values():
    pl = None
    cpnum = '2'
    if mr[0]['player1'] == 'Carlos Alcaraz':
        cpnum = '1'
    for r in mr:
        if pl is None:
            pl = r['point_victor']
            continue
        # If Carlos won current point
        if r['point_victor'] == cpnum:
            # inc sum to indicate carlos won current point after winning the previous one
            p_consec_sum += 1 
        # Increment sum of total points
        p_sum += 1

        pl = r['point_victor']
        
print(p_sum, p_consec_sum, p_consec_sum/p_sum)

1202 639 0.5316139767054908


### Results:
Carlos Alcaraz has a 54.47% chance of winning the next point given that he has won the previous

Carlos Alcaraz has a 51.68% chance of winning the next point given that he has lost the previous

Carlos Alcaraz has a 53.16% chance of winning any given point